src/puppetlabs/puppetdb/command.clj

```diff
@@ -86,23 +101,78 @@
    "store report" (version-range 3 6)
    "deactivate node" (version-range 1 3)})
 
-;; ## Command parsing
+(defn- read-command-body
+  [body]
+  (if (= (class body) (class (byte-array 0)))
+    (json/parse-string (String. body "UTF-8") true)
+    (json/parse-string body true)))
+
+(defn- die-on-header-payload-mismatch
+  [name in-header in-body]
+  (when (and in-header in-body (not= in-header in-body))
+    (throw+
+     (fatality
+      (Exception.
+       (format "%s mismatch between message properties and payload (%s != %s)"
+               name in-header in-body))))))
+
+(defn parse-new-command
+  "Parses a new-format queue command and returns it in the traditional
+  format."
+  [{:keys [headers body]}]
+  ;; New-style messages must have all 5 headers, and the body must be
+  ;; the bare command content (i.e. the payload).  So for
+  ;; a "deactivate node" command,
+  ;; a string like {"certname":"test1","producer_timestamp":"2015-01-01"}.
+  (let [{:keys [command version certname received id]} headers
+        version (if-let [v (:version headers)] (Integer/parseInt v) nil)]
+    (when-not (and command version certname received id)
+      (throw+
+       (fatality (Exception. (str "Missing command message properties:"
+                                  (when-not command " command")
+                                  (when-not version " version")
+                                  (when-not certname " certname")
+                                  (when-not received " received")
+                                  (when-not id " id"))))))
+    (let [message (read-command-body body)]
+      (die-on-header-payload-mismatch "certname"
+                                      certname
+                                      (get-in message ["payload" "certname"]))
+      (assert (not (:annotations message)))
+      {:command command
+       :version version
+       :certname certname ;; Duplicated with respect to payload for now.
+       :payload message
+       :annotations (dissoc headers :command :version :certname)})))
+
+(defn- parse-queue-command
+  [{:keys [headers body]}]
+  ;; Queue messages must not have command, version, or certname headers,
+  ;; and the body must be a JSON map like this:
+  ;; {"command":"deactivate node","version":3, "payload":{"certname":...}}
+  (let [{:keys [command version certname]} headers]
+    (when (or command version certname)
+      (throw+
+       (fatality (Exception. (str "Unexpected command message properties:"
+                                  (when command " command")
+                                  (when version " version")
+                                  (when certname " certname"))))))
+    (update (read-command-body body)
+            :annotations merge
```

Comment:
The way this is line wrapped makes it hard to read... would be easier if `:annotations` was on the previous line, I think.